### PR TITLE
[Refactor] New Halide-Like IR (Remove ISL, YAML deps and Polyhedral Compiler)

### PR DESCRIPTION
**Pull request by hikettei**
### Blueprint New Specifications

- [ ] Caten V2 brings the following feature:
  - [ ] Simplify the entire codebase
  - [ ] Complete AutoScheduler
  - [ ] Improve the compilation speed by refactoring Pattern Matcher
  - [ ] Integrate FastGraph and Graph
  - [ ] Perfect Scheduler in codegen
  - [ ] Complete GPU Support
  - [ ] Llama3 Support
  - [ ] Directly runs on mobile; iOS and Android
- [ ] Gemm >= 300GFLOPS is must
- [ ] Optimize the compilation speed (esp: pattern matcher)
- **Goals**
  - [ ] Bring back ECL, CCL Tests (which was previously failed due to ISL)
  - [x] Blueprint is DAG, Everything is FastGraph
  - [x] Remove weirdness in type inference (remove expr!)
  - [ ] Add X86/PTX Renderer
  - [x] Refactor the entire codebase
  - [x] Use defsimplifier 
  - [x] jit.lisp: tweak the view inference
  - [x] Maybe remove ISL dependencies
  - [x] Tile the range
  - [ ] Rename JIT_DEBUG -> DEBUG
  - [x] Make Everything FastGraph
  - [ ] BEAM Search only
  - [ ] Test Speed in CI !!! (examples/bench.lisp)
- [x] Refactor aasm.lisp
- [x] Various optimizations are applicable after lowering aref indexing
  - [x] TileBands
    - [x] GLOBAL/LOCAL Coincidence
    - [ ] Vectorize
    - [ ] Mapping Matmul+Relu into Microkernel https://arxiv.org/abs/2301.01333 --> clear win!!!
    - [x] TILE
    - [x] Unroll
  - [x] Shared Memory Transfer
    - [x] GROUP
  - [ ] Update the description
  - [ ] Bring Back Memory Planner

```
;; RenderGraph is a DAG
(print-ast
 (with-blueprint
   (%progn
    (%defglobal 'a)
    (%defglobal 'b)
    (%range
     'gid0
     (%progn
      (let ((a (%add (%iconst 1) (%iconst 2)))
            (b (%add (%iconst 2) (%iconst 3))))
        (%mul a b)))
     :start 0 :end (%iconst 'n)))))

[P=0, ID=0]:
   :PROGN {0}
   ├ :DEFINE-GLOBAL {N1}
   ├ :DEFINE-GLOBAL {N2}
   └ :RANGE {N3}
     ├ load(GID0)
     │ └ Allocate[:int64] NIL
     ├ load(N)
     │ └ Allocate[:int64] NIL
     └ :PROGN {N4}
       └ :MUL {N5}
         ├ :ADD {N6}
         │ ├ load(1)
         │ │ └ Allocate[:int64] NIL
         │ └ load(2)
         │   └ Allocate[:int64] NIL
         └ :ADD {N7}
           ├ load(2)
           │ └ Allocate[:int64] NIL
           └ load(3)
             └ Allocate[:int64] NIL
```
- [ ] Implement a subclass of Graph:
  - [ ] RenderGraph
  - [ ] ScheduleGraph
  - [ ] (Normal) TensorGraph
- [x] Rename AASM -> IR
- [x] Move expr -> aasm
- [x] Render OPs definition are in aasm
- [ ] Implement Dequantize using with blueprint
- [ ] CI: Speed test vs PyTorch

### CODE REVIEW

**Comment by hikettei** [comment]
Finally everything fits in air graph, super cool things. Surprisingly I am getting the exactly equivalent schedule w/ ISL Scheduler using rule-based lowerer, which means we can moved the stage of generating optimized kernel for different archs.

Embedding:
```
=== [P: 0] =================================================================================================================================
[P=0, ID=0]:
   :PROGN {0}
   ├ :EXPR {N1}
   │ └ load(0.0)
   │   └ :DEFINE-GLOBAL {N2}
   └ :FOR {N3}
     ├ :RANGE {N4}
     └ :PROGN {N5}
       └ :FOR {N6}
         ├ :RANGE {N7}
         └ :PROGN {N8}
           ├ :EXPR {N9}
           │ └ load(0.0)
           │   └ :AREF {N10}
           │     ├ :DEFINE-GLOBAL {N11}
           │     └ :ADD {N12}
           │       └ :MUL {N13}
           │       │ ├ :RANGE {N14}
           │       │ └ load(128)
           │       │   └ Allocate[:int64] NIL
           │       ├ :ADD {N15}
           │         └ :RANGE {N16}
           │         ├ :MUL {N17}
           │           ├ :RANGE {N4}
           │           └ load(128)
           │             └ Allocate[:int64] NIL
           ├ :FOR {N18}
           │ ├ :RANGE {N14}
           │ └ :PROGN {N19}
           │   └ :EXPR {N20}
           │     └ :ADD {N21}
           │       ├ :EXPR {N9}
           │       └ :WHERE {N22}
           │         └ :EXPR {N1}
           │         ├ :MOVE {N23}
           │         │ └ :AREF {N24}
           │         │ │ ├ :DEFINE-GLOBAL {N25}
           │         │ │ └ :ADD {N26}
           │         │ │   ├ :RANGE {N27}
           │         │ │   └ :MUL {N28}
           │         │ │     └ :RANGE {N29}
           │         │ │     ├ load(128)
           │         │ │       └ Allocate[:int64] NIL
           │         │ ├ :AREF {N30}
           │         │   ├ :DEFINE-GLOBAL {N31}
           │         │   └ :ADD {N32}
           │         │     └ :MUL {N33}
           │         │     │ ├ :RANGE {N14}
           │         │     │ └ load(128)
           │         │     │   └ Allocate[:int64] NIL
           │         │     ├ :ADD {N34}
           │         │       └ :RANGE {N35}
           │         │       ├ :MUL {N36}
           │         │         ├ :RANGE {N4}
           │         │         └ load(16384)
           │         │           └ Allocate[:int64] NIL
           │         ├ :MOVE {N37}
           │           ├ :AREF {N38}
           │           │ ├ :DEFINE-GLOBAL {N39}
           │           │ └ :ADD {N40}
           │           │   └ :MUL {N41}
           │           │   │ ├ :RANGE {N14}
           │           │   │ └ load(128)
           │           │   │   └ Allocate[:int64] NIL
           │           │   ├ :ADD {N42}
           │           │     └ :RANGE {N43}
           │           │     ├ :MUL {N44}
           │           │       ├ :RANGE {N4}
           │           │       └ load(16384)
           │           │         └ Allocate[:int64] NIL
           │           └ :NOT {N45}
           │             └ :!= {N46}
           │               └ :AREF {N47}
           │               │ ├ :DEFINE-GLOBAL {N48}
           │               │ └ :RANGE {N49}
           │               ├ :AREF {N50}
           │               │ ├ :DEFINE-GLOBAL {N51}
           │               │ └ :ADD {N52}
           │               │   └ :RANGE {N53}
           │               │   ├ :ADD {N54}
           │               │     └ :RANGE {N55}
           │               │     ├ :MUL {N56}
           │               │       ├ :RANGE {N4}
           │               │       └ load(128)
           │               │         └ Allocate[:int64] NIL
           │               ├ :MOVE {N57}
           │                 └ :ADD {N58}
           │                 │ ├ :ADD {N59}
           │                 │ └ :RANGE {N60}
           │                 ├ :AREF {N61}
           │                   ├ :DEFINE-GLOBAL {N62}
           │                   └ :ADD {N63}
           │                     └ :RANGE {N64}
           │                     ├ :ADD {N65}
           │                       └ :RANGE {N66}
           │                       ├ :MUL {N67}
           │                         ├ :RANGE {N4}
           │                         └ load(128)
           │                           └ Allocate[:int64] NIL
           └ :EXPR {N68}
             └ :MOVE {N69}
               ├ :AREF {N70}
               │ ├ :DEFINE-GLOBAL {N71}
               │ └ :ADD {N72}
               │   └ :MUL {N73}
               │   │ ├ :RANGE {N14}
               │   │ └ load(128)
               │   │   └ Allocate[:int64] NIL
               │   ├ :ADD {N74}
               │     └ :RANGE {N75}
               │     ├ :MUL {N76}
               │       ├ :RANGE {N4}
               │       └ load(128)
               │         └ Allocate[:int64] NIL
               └ :AREF {N77}
                 └ :ADD {N78}
                 │ └ :RANGE {N79}
                 │ ├ :MUL {N80}
                 │   ├ :RANGE {N4}
                 │   └ load(128)
                 │     └ Allocate[:int64] NIL
                 ├ :EXPR {N20}
============================================================================================================================================
;; Preview
{
  val_17 = 0.0;
  @coincident for (int _gid0=0; _gid0<16384; _gid0+=1) 
  {
    @coincident for (int _gid2=0; _gid2<128; _gid2+=1) 
    {
      val_2 = 0.0;
      @reduction for (int _gid1=0; _gid1<128; _gid1+=1) 
      {
        val_19 = (val_2+!(((_gid1)!=val_9[_gid0])) ? val_14[(_gid2+(128*_gid1))] : val_17);
      }
      val_20 = val_19[((_gid0*128)+_gid2)];
    }
  }
}
```
Softmax:
```
=== [P: 0] =================================================================================================================================
[P=0, ID=0]:
   :PROGN {0}
   ├ :EXPR {N1}
   │ └ load(1.442695)
   │   └ :DEFINE-GLOBAL {N2}
   └ :FOR {N3}
     ├ :RANGE {N4}
     └ :PROGN {N5}
       ├ :EXPR {N6}
       │ └ load(#.SB-EXT:SINGLE-FLOAT-NEGATIVE-INFINITY)
       │   └ :AREF {N7}
       │     ├ :DEFINE-GLOBAL {N8}
       │     └ :ADD {N9}
       │       ├ :RANGE {N10}
       │       └ :RANGE {N11}
       ├ :EXPR {N12}
       │ └ load(0.0)
       │   └ :AREF {N13}
       │     ├ :DEFINE-GLOBAL {N14}
       │     └ :ADD {N15}
       │       ├ :RANGE {N16}
       │       └ :RANGE {N17}
       ├ :FOR {N18}
       │ ├ :RANGE {N19}
       │ └ :PROGN {N20}
       │   └ :EXPR {N21}
       │     └ :MAX {N22}
       │       ├ :EXPR {N6}
       │       └ :AREF {N23}
       │         ├ :EXPR {N24}
       │         │ └ :EXPR {N25}
       │         │   └ :DEFINE-GLOBAL {N26}
       │         └ :ADD {N27}
       │           └ :RANGE {N28}
       │           ├ :MUL {N29}
       │             ├ :RANGE {N4}
       │             └ load(256)
       │               └ Allocate[:int64] NIL
       ├ :FOR {N30}
       │ ├ :RANGE {N19}
       │ └ :PROGN {N31}
       │   ├ :EXPR {N24}
       │   └ :EXPR {N32}
       │     └ :ADD {N33}
       │       ├ :EXPR {N12}
       │       └ :AREF {N34}
       │         └ :ADD {N35}
       │         │ └ :RANGE {N36}
       │         │ ├ :MUL {N37}
       │         │   ├ :RANGE {N4}
       │         │   └ load(256)
       │         │     └ Allocate[:int64] NIL
       │         ├ :EXPR {N38}
       │           └ :EXP2 {N39}
       │             └ :MUL {N40}
       │               └ :EXPR {N1}
       │               ├ :ADD {N41}
       │                 ├ :AREF {N42}
       │                 │ ├ :EXPR {N24}
       │                 │ └ :ADD {N43}
       │                 │   └ :RANGE {N44}
       │                 │   ├ :MUL {N45}
       │                 │     ├ :RANGE {N4}
       │                 │     └ load(256)
       │                 │       └ Allocate[:int64] NIL
       │                 └ :NEG {N46}
       │                   └ :AREF {N47}
       │                     └ :RANGE {N48}
       │                     ├ :EXPR {N21}
       └ :FOR {N49}
         ├ :RANGE {N19}
         └ :PROGN {N50}
           ├ :EXPR {N24}
           ├ :EXPR {N38}
           └ :EXPR {N51}
             └ :MUL {N52}
               ├ :AREF {N53}
               │ └ :ADD {N54}
               │ │ └ :RANGE {N55}
               │ │ ├ :MUL {N56}
               │ │   ├ :RANGE {N4}
               │ │   └ load(256)
               │ │     └ Allocate[:int64] NIL
               │ ├ :EXPR {N38}
               └ :RECIP {N57}
                 └ :AREF {N58}
                   └ :RANGE {N59}
                   ├ :EXPR {N32}
============================================================================================================================================
{
  val_7 = 1.442695;
  @coincident for (int _gid0=0; _gid0<256; _gid0+=1) 
  {
    val_2 = #.sb-ext:single-float-negative-infinity;
    val_11 = 0.0;
    @reduction for (int _gid1=0; _gid1<256; _gid1+=1) 
    {
      val_3 = max(val_2, val_0[((_gid0*256)+_gid1)]);
    }
    @reduction for (int _gid1=0; _gid1<256; _gid1+=1) 
    {
      val_0 = e1632534;
      val_12 = (val_11+val_9[((_gid0*256)+_gid1)]);
    }
    @coincident for (int _gid1=0; _gid1<256; _gid1+=1) 
    {
      val_9 = exp2(((val_0[((_gid0*256)+_gid1)]+-(val_3[_gid0]))*val_7));
      val_14 = (val_9[((_gid0*256)+_gid1)]*1/(val_12[_gid0]));
    }
  }
}
```

**Comment by hikettei** [comment]
Matmul Scheduling (MaximizeBand)
```
 ============================================================================================================================================
=== [P: 0] =================================================================================================================================
[P=0, ID=0]:
   :PROGN {0}
   └ :FOR {N1}
     ├ :RANGE {N2}
     │ ├ :EXPR {N3}
     │   └ load(M)
     │     └ Allocate[:int64] NIL
     └ :FOR {N4}
       ├ :RANGE {N5}
       │ ├ :EXPR {N6}
       │   └ load(K)
       │     └ Allocate[:int64] NIL
       └ :PROGN {N7}
         ├ :EXPR {N8}
         │ └ load(0.0)
         │   └ :AREF {N9}
         │     ├ :DEFINE-GLOBAL {N10}
         │     └ :ADD {N11}
         │       └ :RANGE {N12}
         │       │ ├ :EXPR {N13}
         │       │   └ load(N)
         │       │     └ Allocate[:int64] NIL
         │       ├ :ADD {N14}
         │         └ :RANGE {N15}
         │         │ ├ :EXPR {N6}
         │         ├ :MUL {N16}
         │           └ load(K)
         │           │ └ Allocate[:int64] NIL
         │           ├ :RANGE {N2}
         ├ :FOR {N17}
         │ ├ :RANGE {N18}
         │ │ ├ :EXPR {N13}
         │ └ :EXPR {N19}
         │   └ :ADD {N20}
         │     ├ :EXPR {N8}
         │     └ :MUL {N21}
         │       └ :AREF {N22}
         │       │ ├ :DEFINE-GLOBAL {N23}
         │       │ └ :ADD {N24}
         │       │   ├ :RANGE {N25}
         │       │   │ ├ :EXPR {N6}
         │       │   └ :MUL {N26}
         │       │     ├ load(K)
         │       │     │ └ Allocate[:int64] NIL
         │       │     └ :RANGE {N27}
         │       │       ├ :EXPR {N13}
         │       ├ :MOVE {N28}
         │         └ :AREF {N29}
         │         │ ├ :DEFINE-GLOBAL {N30}
         │         │ └ :ADD {N31}
         │         │   └ :RANGE {N32}
         │         │   │ ├ :EXPR {N13}
         │         │   ├ :MUL {N33}
         │         │     ├ load(N)
         │         │     │ └ Allocate[:int64] NIL
         │         │     └ :RANGE {N34}
         │         │       ├ :EXPR {N3}
         │         ├ :AREF {N35}
         │           ├ :DEFINE-GLOBAL {N36}
         │           └ :ADD {N37}
         │             └ :RANGE {N38}
         │             │ ├ :EXPR {N13}
         │             ├ :ADD {N39}
         │               └ :MUL {N40}
         │               │ └ load(N)
         │               │ │ └ Allocate[:int64] NIL
         │               │ ├ :RANGE {N5}
         │               ├ :MUL {N41}
         │                 ├ :RANGE {N2}
         │                 └ :MUL {N42}
         │                   ├ load(N)
         │                   │ └ Allocate[:int64] NIL
         │                   └ load(K)
         │                     └ Allocate[:int64] NIL
         └ :EXPR {N43}
           └ :MOVE {N44}
             ├ :AREF {N45}
             │ ├ :DEFINE-GLOBAL {N46}
             │ └ :ADD {N47}
             │   └ :RANGE {N48}
             │   │ ├ :EXPR {N13}
             │   ├ :ADD {N49}
             │     └ :RANGE {N50}
             │     │ ├ :EXPR {N6}
             │     ├ :MUL {N51}
             │       └ load(K)
             │       │ └ Allocate[:int64] NIL
             │       ├ :RANGE {N2}
             └ :AREF {N52}
               └ :ADD {N53}
               │ └ :RANGE {N54}
               │ │ ├ :EXPR {N6}
               │ ├ :MUL {N55}
               │   ├ load(K)
               │   │ └ Allocate[:int64] NIL
               │   └ :RANGE {N56}
               │     ├ :EXPR {N3}
               ├ :EXPR {N19}
============================================================================================================================================
{
  @coincident for (int _gid0=0; _gid0<m; _gid0+=1)  [B1]
    @coincident for (int _gid1=0; _gid1<k; _gid1+=1)  [B1]
    {
      val_7 = 0.0; // expr
      @reduction for (int _gid2=0; _gid2<n; _gid2+=1) 
        val_21 = (val_7+(val_16[((n*_gid0)+_gid2)]*val_19[(_gid1+(k*_gid2))])); // expr
      val_22 = val_21[((k*_gid0)+_gid1)]; // expr
    }
}
```
